pkg/pillar/types/global.go

```diff
@@ -103,312 +109,439 @@ func (gs *GlobalStatus) UpdateItemValuesFromGlobalConfig(gc GlobalConfig) {
+}
+
+func (configPtr *ConfigItemValueMap) SetAgentValue(key AgentSettingKey, agentName string, newValue string) error {
+	if configPtr.AgentSettings == nil {
+		configPtr.AgentSettings = make(map[string]map[AgentSettingKey]ConfigItemValue)
 	}
-	if newgc.NetworkTestBetterInterval < GlobalConfigMinimums.NetworkTestBetterInterval {
-		log.Warnf("Enforce minimum NetworkTestInterval received %d; using %d",
-			newgc.NetworkTestBetterInterval, GlobalConfigMinimums.NetworkTestBetterInterval)
-		newgc.NetworkTestBetterInterval = GlobalConfigMinimums.NetworkTestBetterInterval
+	if settingMap, ok := configPtr.AgentSettings[agentName]; ok {
+		// This can double as a create. If the key is not found then it creates it anyway
+		settingMap[key] = ConfigItemValue{stringValue: newValue}
+	} else {
+		return fmt.Errorf("unable to modify per-agent setting for agent: %s because agent was not found", agentName)
 	}
+	return nil
+}
 
-	if newgc.StaleConfigTime < GlobalConfigMinimums.StaleConfigTime {
-		log.Warnf("Enforce minimum StaleConfigTime received %d; using %d",
-			newgc.StaleConfigTime, GlobalConfigMinimums.StaleConfigTime)
-		newgc.StaleConfigTime = GlobalConfigMinimums.StaleConfigTime
+func (configPtr *ConfigItemValueMap) SetGlobalValue(key GlobalSettingKey, newValue string) error {
+	if configPtr.GlobalSettings == nil {
+		configPtr.GlobalSettings = make(map[GlobalSettingKey]ConfigItemValue)
 	}
-	if newgc.DownloadGCTime < GlobalConfigMinimums.DownloadGCTime {
-		log.Warnf("Enforce minimum DownloadGCTime received %d; using %d",
-			newgc.DownloadGCTime, GlobalConfigMinimums.DownloadGCTime)
-		newgc.DownloadGCTime = GlobalConfigMinimums.DownloadGCTime
+	if _, ok := configPtr.GlobalSettings[key]; ok {
+		configPtr.GlobalSettings[key] = ConfigItemValue{stringValue: newValue}
+	} else {
+		return fmt.Errorf("unable to modify global setting %s because setting was not found", string(key))
 	}
-	if newgc.VdiskGCTime < GlobalConfigMinimums.VdiskGCTime {
-		log.Warnf("Enforce minimum VdiskGCTime received %d; using %d",
-			newgc.VdiskGCTime, GlobalConfigMinimums.VdiskGCTime)
-		newgc.VdiskGCTime = GlobalConfigMinimums.VdiskGCTime
+	return nil
+}
+
+func (configPtr *ConfigItemValueMap) DelAgentValue(key AgentSettingKey, agentName string) error {
+	// I don't think this is needed because if you are calling delete, the key must already exist. But maybe just in case the user makes a mistake?
+	if configPtr.AgentSettings == nil {
+		configPtr.AgentSettings = make(map[string]map[AgentSettingKey]ConfigItemValue)
+	}
+	if settingMap, ok := configPtr.AgentSettings[agentName]; ok {
+		if _, okVal := settingMap[key]; okVal {
+			delete(settingMap, key)
+		} else {
+			return fmt.Errorf("unable to delete per-agent setting for agent: %s because setting %s was not found",
+				agentName, string(key))
+		}
+	} else {
+		return fmt.Errorf("unable to delete per-agent setting for agent: %s because agent was not found",
+			agentName)
+	}
+	return nil
```

Comment:
This function should really be:
   settingMap, ok := configPtr.AgentSettings[agentName]
   if !ok {
      return
   }
   delete(settingMap, key)
   if len(settingMap) > 0 {
      configPtr.AgentSettings[agentName] = settingMap
   } else {
      // No more settings for Agent.. So delete it from AgentSettings
      delete(configPtr.AgentSettings, agentName)
   }
 }